mods/default/workbench.lua

```diff
@@ -0,0 +1,234 @@
+				return stack:get_count()
+			end
+		elseif listname == "output" then
+			return 0
+		elseif listname == "input" then
+			if minetest.registered_items[stack:get_name()].groups.renameable == 0 then
+				return 0
+			end
+		end
+
+		return 1
+	end,
+	allow_metadata_inventory_move = function(pos, from_list, from_index, to_list, to_index, count, player)
+		local inv = minetest.get_meta(pos):get_inventory()
+		local stack = inv:get_stack(from_list, from_index)
+
+		if to_list == "nametag" then
+			if stack:get_name() ~= "default:nametag" then
+				return 0
+			end
+		elseif to_list == "input" then
+			if minetest.registered_items[stack:get_name()].groups.renameable == 0 then
+				return 0
+			end
+		elseif to_list == "output" then
+			return 0
+		end
+
+		return count
+	end,
+	on_metadata_inventory_put = function(pos, listname, index, stack)
+		local meta = minetest.get_meta(pos)
+		local inv = meta:get_inventory()
+		local error = meta:get_string("error")
+
+		if error == "input" and not inv:is_empty("input") then
+			meta:set_string("formspec", get_workbench_formspec(pos))
+		elseif error == "nametag" and not inv:is_empty("nametag") then
+			meta:set_string("formspec", get_workbench_formspec(pos))
+		end
+
+		if listname == "input" then
+			workbench_update_text(pos, stack)
+		end
+	end,
+	on_metadata_inventory_move = function(pos, from_list, from_index, to_list, to_index)
+		local meta = minetest.get_meta(pos)
+		local inv = meta:get_inventory()
+		local stack = inv:get_stack(to_list, to_index)
+
+		if to_list == "input" then
+			workbench_update_text(pos, stack)
+		end
+	end,
+	on_metadata_inventory_take = function(pos, listname)
+		local meta = minetest.get_meta(pos)
+		local inv = meta:get_inventory()
+		local error = meta:get_string("error")
+
+		if error == "output" and inv:is_empty("output") then
+			meta:set_string("formspec", get_workbench_formspec(pos))
+		end
+
+		if listname == "input" then
+			meta:set_string("text", "")
+			meta:set_string("formspec", get_workbench_formspec(pos))
+		end
+	end,
+	on_receive_fields = function(pos, formname, fields, sender)
+		local meta = minetest.get_meta(pos)
+		local inv = meta:get_inventory()
+
+		if fields.rename or fields.key_enter_field == "text" then
+			meta:set_string("text", fields.text)
+
+			if inv:is_empty("input") then
+				workbench_update_help(pos, "input", "Missing input item!")
+			elseif inv:is_empty("nametag") then
+				workbench_update_help(pos, "nametag", "Missing nametag!")
+			elseif not inv:is_empty("output") then
+				workbench_update_help(pos, "output", "No room in output!")
+			else
+				if fields.text == "" then
+					workbench_update_help(pos, nil, "Description cannot be blank!")
+					return
+				elseif fields.text:len() > 500 then
+					workbench_update_help(pos, nil, "Description too long (max 500 characters)!")
+					return
+				elseif fields.text == get_item_desc(inv:get_stack("input", 1)) then
+					workbench_update_help(pos, nil, "Description not changed!")
```

Comment:
I inserted this check as I thought that it'd be useless to "rename" an item if the new description was the same as the current description, therefore wasting a nametag. I can see that it could be confusing though, and maybe should be allowed.